src/main/java/org/truffleruby/core/queue/SizedQueueNodes.java

```diff
@@ -182,8 +178,36 @@ private Object doPop(SizedQueue queue) {
             return getContext().getThreadManager().runUntilResult(this, queue::take);
         }
 
+        @Specialization(guards = "!nonBlocking")
+        protected Object popBlocking(RubySizedQueue self, boolean nonBlocking, long timeoutMilliseconds) {
+            final SizedQueue queue = self.queue;
+            final long deadline = System.currentTimeMillis() + timeoutMilliseconds;
+
+            var result = getContext().getThreadManager().runUntilResult(this, () -> {
+                final long currentTimeout = deadline - System.currentTimeMillis();
+                final Object value;
+
+                if (currentTimeout > 0) {
+                    value = queue.poll(currentTimeout);
+                } else {
+                    value = queue.poll();
+                }
+
+                if (value == SizedQueue.CLOSED) {
+                    throw new RaiseException(getContext(), coreExceptions().closedQueueError(this));
+                } else {
+                    return value;
+                }
+            });
```

Comment:
Wondering how a case with expired timeout is handled. Could we debug this case?

When a blocking `pop` call with timeout happens and timeout expires then I suppose the following things happen (please correct me if I am wrong):
1. `canTake.await` returns `false` in `SizedQueue.poll`
2. `SizedQueue.poll` returns `null`
3. `null` is returned from a lambda, passed as argument to the `runUntilResult` method
4. `runUntilResult` re-runs the lambda as far as it's waiting for the first non-`null` result
5. as far as timeout is expired then `currentTimeout > 0` is `false` and `queue.poll()` is called
6. `queue.poll()` returns `null` for an empty queue (our queue is still empty as far as `poll` time outed)
7. and `runUntilResult` re-runs the lambda again and again until `queue.poll()`  returns non-`null` value

So this way a blocking `pop` call with timeout when a queue is empty will wait longer than the specified timeout -  until some value is added to the queue.

Reply:
My bad. It seems `runUntilResult` doesn't wait for the first non-null value.